### DISCUSSION TITLE
AWS EC2 > Serverpod > Active | Upload

### DISCUSSION THREAD

**Discussion by AndryHTC** [serverpod/serverpod#459] [Q&A] [answered]
Hi...
I'm going to debug a [Prisma](https://github.com/odroe/prisma-dart) issue with the production env https://github.com/odroe/prisma-dart/issues/74 

There is a need for a little bit of explanation of how the deployed EC2 Server directory is composed. Inside the Serverpod folder at the root of the deployment, there are `active `and `upload`, which seem to be 2 identical folders with the relatives `serverpod_server` and `vendor` folders.

What is the explanation for those? And which one do I need to touch for debugging the actual running Serverpod instance?

**Comment by yahu1031** [comment]
> there are `active` and `upload`, which seem to be 2 identical folders with the relatives `serverpod_server` and `vendor` folders.

@AndryHTC Can you elaborate a little bit about it? Or point us at what exactly you mean `upload` and `active`

**Reply by AmoonPod** [reply]
As @AndryHTC says
> Inside the Serverpod folder at the root of the deployment

**Reply by yahu1031** [reply]
@AndryHTC They look like coming from [appspec.yaml](https://docs.aws.amazon.com/codedeploy/latest/userguide/application-specification-files.html) from `PROJECTNAME_server/aws/scripts/appspec.yml`.

As of the scripts written, the `PROJECTNAME_server` is the one the current deployment is going in. And `vendor` is a kind of backup IG.

**Reply by AndryHTC** [reply]
> As of the scripts written, the `PROJECTNAME_server` is the one the current deployment is going in. And `vendor` is a kind of backup IG.

@yahu1031 thank you for the efforts, but `PROJECTNAME_server` and `vendor` are not the problems... 

1. As described in the documentation, `vendor` should contain [External dependencies and submodules](https://docs.serverpod.dev/deployments/deploying-to-aws#external-dependencies-and-submodules)
2. The problem is that these folders are _**both**_ contained by **_both_** `active` and `upload` folders, and so I don't understand what the purpose of them

**Reply by yahu1031** [reply]
@AndryHTC I Found something in the [`start_server`](https://github.com/serverpod/serverpod/blob/70473726427c4989ef0506473c08412a97cff80e/templates/serverpod_templates/projectname_server/aws/scripts/start_server#L16-L17) scripts again. The command in line 17 copies the whole sub-directories and their attributes to the `active` folder/directory. Maybe I guess, an `active` directory is being used in deployment. where the `upload` holds the backup of it. If any changes wanna be made, must be done in the [`active` directory to get affected in the server](https://github.com/serverpod/serverpod/blob/70473726427c4989ef0506473c08412a97cff80e/templates/serverpod_templates/projectname_server/aws/scripts/run_serverpod#L4).

**Reply by yahu1031** [reply]
Or change the things in the `upload` folder and run the `start_server` it will copy the updated code and reflect it.

**Comment by vlidholt** [comment, marked as answer]
The deployment process from Github works the following way:

1. The deployment is compiled on Github, then uploaded to a bucket on S3.
2. CodeDeploy is used to copy the deployment to the upload folder on the EC2 instance.
3. The running server is stopped, the `upload` directory is copied to the `active` directory, then the new server is started.

This is done to minimize any downtime.

As for the `vendor` directory. This is where you put any external dependencies (which are not loaded from pub.dev). For instance, if you use Serverpod as a submodule, this is the place for it. The `vendor` directory will also be bundled with the deployment.
